### PR TITLE
test: introduce TEST-08-INITRD

### PR DESCRIPTION
**Pull request by mrc0mmand**
And move the initrd related tests from TEST-01-BASIC there.

Additionally, this should provide coverage for recemt shutdown initrd
related issues, see:
  - https://github.com/systemd/systemd/issues/28645
  - https://github.com/systemd/systemd/pull/28648
  - https://github.com/systemd/systemd/pull/28793

### CODE REVIEW

**Comment by mrc0mmand** [comment]
I just realized that TEST-01-BASIC now contains tests for both "regular" and shutdown initrd, so maybe it would be better to split that stuff to a separate TEST-XX-INITRD test to make the TEST-01-BASIC test basic again.

File: test/TEST-08-INITRD/test.sh [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        derror "Missing 'hello' message from shutdown initrd"
+        return 1
+    fi
+
+    return 0
+}
+
+# Setup a one shot service in initrd that creates a dummy bind mount under /run
+# to check if the mount persists though the initrd transition. The "check" part
+# is in the respective testsuite-01.sh script.
```
Comment (yuwata) [inline test/TEST-08-INITRD/test.sh]: 01 -> 08

File: test/TEST-08-INITRD/test.sh [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    mkdir -m 755 "$extra/etc" "$extra/etc/systemd" "$extra/etc/systemd/system" "$extra/etc/systemd/system/initrd.target.wants"
+
+    cat >"$extra/etc/systemd/system/initrd-run-mount.service" <<EOF
+[Unit]
+Description=Create a mount in /run that should survive the transition from initrd
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=sh -xec "mkdir /run/initrd-mount-source /run/initrd-mount-target; mount -v --bind /run/initrd-mount-source /run/initrd-mount-target"
```
Comment (yuwata) [inline test/TEST-08-INITRD/test.sh]: I think it is better to touch a file in the mount point, and check the existence after switching root.

**Review by yuwata — APPROVED** [review]
Looks nice!

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
Hmm.
```
$ sudo make -C TEST-08-INITRD BUILD_DIR=/var/build clean setup run
(snip)
I: Generate modules dependencies
depmod: WARNING: could not open modules.builtin.modinfo at /var/tmp/systemd-tests/systemd-test.rGaua6/root/lib/modules/6.3.12-200.fc38.x86_64: No such file or directory
Reusing existing cached image /var/build/test/TEST-08-INITRD/../default.img → /var/build/test/default.img
'/var/build/test/default.img' -> '/var/tmp/systemd-tests/systemd-test.rGaua6/default.img'
I: Masking supporting services
'/var/tmp/systemd-tests/systemd-test.rGaua6/root/etc/systemd/system/systemd-hwdb-update.service' -> '/dev/null'
'/var/tmp/systemd-tests/systemd-test.rGaua6/root/etc/systemd/system/systemd-journal-catalog-update.service' -> '/dev/null'
'/var/tmp/systemd-tests/systemd-test.rGaua6/root/etc/systemd/system/systemd-networkd.service' -> '/dev/null'
'/var/tmp/systemd-tests/systemd-test.rGaua6/root/etc/systemd/system/systemd-networkd.socket' -> '/dev/null'
'/var/tmp/systemd-tests/systemd-test.rGaua6/root/etc/systemd/system/systemd-resolved.service' -> '/dev/null'
cp: cannot create regular file '/var/tmp/systemd-tests/systemd-test.rGaua6/root/shutdown-initrd//usr/lib64/systemd/libsystemd-shared-254.so': No such file or directory
make: *** [Makefile:4: setup] Error 1
make: Leaving directory '/home/watanabe/git/systemd/test/TEST-08-INITRD'
```

File: test/TEST-08-INITRD/test.sh [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    # This should provide coverage for shutdown initrd related issues, see:
+    #   - https://github.com/systemd/systemd/issues/28645
+    #   - https://github.com/systemd/systemd/pull/28648
+    #   - https://github.com/systemd/systemd/pull/28793
+    #
+    # This is a bit messier than I originally anticipated, as installing our own libraries
+    # is handled implicitly by install_systemd() which I don't want to use here, since
+    # I need only the systemd-shutdown binary
+    sd_initrd="$workspace/shutdown-initrd"
+    mkdir -p "$sd_initrd/etc" "$sd_initrd/usr/lib/systemd" "$sd_initrd/usr/lib64"
```
Comment (yuwata) [inline test/TEST-08-INITRD/test.sh]: Changing
`"$sd_initrd/usr/lib64"` -> `"$sd_initrd/usr/lib64/systemd"`
fixes my issue.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: test/TEST-08-INITRD/test.sh [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    initdir="$sd_initrd" image_install bash /usr/lib/os-release
+    ln -srf "$sd_initrd/usr/lib/os-release" "$sd_initrd/etc/initrd-release"
+    cp -u "$workspace/usr/lib/systemd/systemd-shutdown" "$sd_initrd/usr/lib/systemd/systemd-shutdown"
+    initdir="$sd_initrd" inst_libs "$sd_initrd/usr/lib/systemd/systemd-shutdown"
+    # We need to deal with libsystemd stuff explicitly, as we don't call install_systemd() here
+    while read -r file; do
+        # /.../workspace/usr/lib64/systemd/foo.so => /usr/lib64/systemd
+        dir="$(dirname "${file##"$workspace"}")"
+        mkdir -p "${dir:?}"
+        cp -u "$file" "${sd_initrd:?}/${file##"$workspace"}"
```
Comment (yuwata) [inline test/TEST-08-INITRD/test.sh]: I think, better way is changing the line to inst_library (or inst_simple)
```
initdir="$sd_initrd" inst_library "$file" "${file##"$workspace"}"
```

**Review by yuwata — CHANGES_REQUESTED** [review]

File: test/TEST-08-INITRD/test.sh (hunk 4 above)
Comment (mrc0mmand) [reply test/TEST-08-INITRD/test.sh]: Thank you! That makes the code look less like scrambled eggs.

**Review by yuwata — APPROVED** [review]

**Review by bluca — APPROVED** [review]

**Comment by bluca** [comment]
@yuwata the new test is failing on Ubuntu, did you miss that?

```
6007s [   23.982135] testsuite-08.sh[351]: + test -d /run/initrd-mount-target
6007s [ESC[0;1;31mFAILEDESC[0m] Failed to start ESC[0;1;39mTEST-08-INITRDESC[0m.
```

**Comment by yuwata** [comment]
Ouch, these CIs also fail on TEST-58-REPART, so I overlooked the failure in TEST-08-INITRD, sorry.

**Comment by yuwata** [comment]
```
5981s Booting from ROM...
5987s c[?7l[2J[    4.189902] Initramfs unpacking failed: Decoding failed     <--- what?
5989s Loading, please wait...
5989s Starting systemd-udevd version 254.519.gb6c51fff1b.0
```

**Comment by bluca** [comment]
I think the custom initrd test never worked on distros that don't use dracut. Previously it was just skipped, as under debian/ubuntu TEST-01 runs in nspawn, so it was never exercides.

**Comment by mrc0mmand** [comment]
Can you temporarily blacklist the test on Ubuntu CI? I am currently afk, but once I get home I'll look into some way to make the test work everywhere.

**Comment by yuwata** [comment]
Sure. I will disable it soon.

**Comment by mrc0mmand** [comment]
> I think the custom initrd test never worked on distros that don't use dracut. Previously it was just skipped, as under debian/ubuntu TEST-01 runs in nspawn, so it was never exercides.

Ugh. So, the "decoding failed" message is actually a red herring - the custom initrd works fine (since it's always a cpio archive no matter the distribution). However, Ubuntu (and I guess Debian?) apparently doesn't run systemd in initrd, only udev:

```
# lsinitramfs /boot/initrd.img | grep systemd
usr/lib/modprobe.d/systemd.conf
usr/lib/systemd
usr/lib/systemd/network
usr/lib/systemd/network/73-usb-net-by-mac.link
usr/lib/systemd/network/99-default.link
usr/lib/systemd/systemd-udevd
usr/lib/x86_64-linux-gnu/libsystemd.so.0
usr/lib/x86_64-linux-gnu/libsystemd.so.0.32.0
```

So, even though the necessary files are there:

```
(initramfs) find /etc/systemd
/etc/systemd
/etc/systemd/system
/etc/systemd/system/initrd-run-mount.service
/etc/systemd/system/initrd.target.wants
/etc/systemd/system/initrd.target.wants/initrd-run-mount.service
```

there's no systemd to actually do the stuff we need:
```
(initramfs) find /usr -name "systemd*"
/usr/lib/systemd
/usr/lib/systemd/systemd-udevd
/usr/lib/modprobe.d/systemd.conf
(initramfs) ps | grep systemd
  136 root     11996 S    /lib/systemd/systemd-udevd --daemon --resolve-names=
  237 root      2932 R    {exe} grep systemd
```

**Comment by bluca** [comment]
That's right, it's shell scripts all the way down

**Pull request merged**